packages/core/src/rules/common/component-name-unique.ts

```diff
@@ -0,0 +1,158 @@
+  Oas3Schema,
+  OasRef,
+} from '../../typings/openapi';
+import { Oas2Definition } from '../../typings/swagger';
+
+const TYPE_NAME_SCHEMA = 'Schema';
+const TYPE_NAME_PARAMETER = 'Parameter';
+const TYPE_NAME_RESPONSE = 'Response';
+const TYPE_NAME_REQUEST_BODY = 'RequestBody';
+
+const TYPE_NAME_TO_OPTION_NAME: { [key: string]: string } = {
+  [TYPE_NAME_SCHEMA]: 'schema',
+  [TYPE_NAME_PARAMETER]: 'parameter',
+  [TYPE_NAME_RESPONSE]: 'response',
+  [TYPE_NAME_REQUEST_BODY]: 'requestBody',
+};
+
+export const ComponentNameUnique: Oas3Rule | Oas2Rule = (options) => {
+  const components = new Map<string, Set<string>>();
+
+  const typeNames: string[] = [];
+  if (options.schema !== 'off') {
+    typeNames.push(TYPE_NAME_SCHEMA);
+  }
+  if (options.parameter !== 'off') {
+    typeNames.push(TYPE_NAME_PARAMETER);
+  }
+  if (options.response !== 'off') {
+    typeNames.push(TYPE_NAME_RESPONSE);
+  }
+  if (options.requestBody !== 'off') {
+    typeNames.push(TYPE_NAME_REQUEST_BODY);
+  }
+
+  function getOptionNameForTypeName(typeName: string): string | null {
+    return TYPE_NAME_TO_OPTION_NAME[typeName] ?? null;
+  }
+
+  function getComponentNameFromAbsoluteLocation(absoluteLocation: string): string {
+    const componentName = absoluteLocation.split('/').slice(-1)[0];
+    if (
+      componentName.endsWith('.yml') ||
+      componentName.endsWith('.yaml') ||
+      componentName.endsWith('.json')
+    ) {
+      return componentName.slice(0, componentName.lastIndexOf('.'));
+    }
+    return componentName;
+  }
+
+  function addFoundComponent(
+    typeName: string,
+    componentName: string,
+    absoluteLocation: string
+  ): void {
+    const key = getKeyForComponent(typeName, componentName);
+    const locations = components.get(key) ?? new Set();
+    locations.add(absoluteLocation);
+    components.set(key, locations);
+  }
+
+  function getKeyForComponent(typeName: string, componentName: string): string {
+    return `${typeName}/${componentName}`;
+  }
+  function getComponentFromKey(key: string): { typeName: string; componentName: string } {
+    const [typeName, componentName] = key.split('/');
+    return { typeName, componentName };
+  }
+
+  function addComponentFromAbsoluteLocation(typeName: string, absoluteLocation: string): void {
+    const componentName = getComponentNameFromAbsoluteLocation(absoluteLocation);
+    addFoundComponent(typeName, componentName, absoluteLocation);
+  }
+
+  const rule: Oas3Visitor = {
+    ref: {
+      leave(ref: OasRef, { type, resolve }: UserContext) {
+        const typeName = type.name;
+        if (typeNames.includes(typeName)) {
+          const resolvedRef = resolve(ref);
+          if (!resolvedRef.location) return;
+
+          addComponentFromAbsoluteLocation(
+            typeName,
+            resolvedRef.location.absolutePointer.toString()
+          );
+        }
+      },
+    },
+    Root: {
+      leave(root: Oas3Definition | Oas2Definition, ctx: UserContext) {
```

Comment:
I noticed inconsistency with [this line](https://github.com/Redocly/redocly-cli/pull/1134/files#diff-e086c7501a57ca548a3dfb56a69a297d4733596de3b0da7cf8ebf6080f25bb53R82). Shouldn't it be like this: 
```suggestion
      leave(root: Oas3Definition, ctx: UserContext) {
```

Reply:
Done 👍